### PR TITLE
Add downtime detector module

### PR DESCRIPTION
**Pull request by ValarDragon**
Cref #3603 

## What is the purpose of the change

Adds downtime detector module.

- Remaining elements at time of PR open
  - [x] Tests
    - [x] Keeper methods
    - [x] CLI
    - [x] Query
    - [x] types
  - [x] app.go wiring

## Brief Changelog

- Adds downtime detector module
  - Change downtime to an enum
  - Implement begin block logic
  - Implement initialization logic
  - Implement query 
  - Implement CLI logic (pending #3647 )

## Testing and Verifying

TODO

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: app/apptesting/test_suite.go [hunk 1]
```diff
@@ -374,7 +374,7 @@ func (s *KeeperTestHelper) BuildTx(
 // StateNotAltered validates that app state is not altered. Fails if it is.
 func (s *KeeperTestHelper) StateNotAltered() {
 	oldState := s.App.ExportState(s.Ctx)
-	s.Commit()
+	s.App.Commit()
```
Comment (ValarDragon) [inline app/apptesting/test_suite.go:377]: This was needed to fix "QueryStateAltered" tests, because Commit would start a new block, which would cause a state write from downtime detector.

File: app/keepers/modules.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ import (
 	ica "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts"
 
 	_ "github.com/osmosis-labs/osmosis/v13/client/docs/statik"
+	downtimemodule "github.com/osmosis-labs/osmosis/v13/x/downtime-detector/downtimedetector_module"
```
Comment (nicolaslara) [inline app/keepers/modules.go:31]: nit: `downtime-detector/downtimedetector_module` is ugly. Can it be `downtime-detector/module`? or `sdkmodule`, or something similar?

File: proto/osmosis/downtime-detector/v1beta1/downtime_duration.proto [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import "gogoproto/gogo.proto";
+import "google/protobuf/any.proto";
+import "cosmos_proto/cosmos.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/osmosis-labs/osmosis/v13/x/downtime-detector/types";
+
+enum Downtime {
```
Comment (nicolaslara) [inline proto/osmosis/downtime-detector/v1beta1/downtime_duration.proto:12]: I noticed you're not using `iota`. Is that so that we don't accidentally change the internal representation by reordering these?

File: x/downtime-detector/abci.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+func (k *Keeper) BeginBlock(ctx sdk.Context) {
+	curTime := ctx.BlockTime()
+	lastBlockTime, err := k.GetLastBlockTime(ctx)
+	if err != nil {
+		ctx.Logger().Error("Downtime-detector, could not get last block time, did initialization happen correctly. " + err.Error())
+	}
+	downtime := curTime.Sub(lastBlockTime)
+	if downtime > 30*time.Second {
```
Comment (nicolaslara) [inline x/downtime-detector/abci.go:18]: nit: shouldn't we just do `k.saveDowntimeUpdates(ctx, downtime)` here and have that function determine when to save the downtime or not ?

File: x/downtime-detector/abci.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+		k.saveDowntimeUpdates(ctx, downtime)
+	}
+	k.StoreLastBlockTime(ctx, curTime)
+}
+
+func (k *Keeper) saveDowntimeUpdates(ctx sdk.Context, downtime time.Duration) {
```
Comment (nicolaslara) [inline x/downtime-detector/abci.go:24]: can we get a gdoc here?

File: x/superfluid/client/cli/query_test.go [hunk 6]
```diff
@@ -90,12 +90,14 @@ func (s *QueryTestSuite) TestQueriesNeverAlterState() {
...
 			&types.ConnectedIntermediaryAccountResponse{},
 		},
-		{
-			"Query estimate sfs delegated amount by validator & denom",
-			"/osmosis.superfluid.Query/EstimateSuperfluidDelegatedAmountByValidatorDenom",
-			&types.EstimateSuperfluidDelegatedAmountByValidatorDenomRequest{ValidatorAddress: s.val.String(), Denom: "gamm/pool/1"},
-			&types.EstimateSuperfluidDelegatedAmountByValidatorDenomResponse{},
-		},
+		// need to adapt s.val.String() to have an intermediate account,
+		// else the response is nil and theres a panic internally.
```
Comment (nicolaslara) [inline x/superfluid/client/cli/query_test.go:94]: Is this related to the downtime detector?

File: go.mod [hunk 7]
```diff
@@ -5,7 +5,7 @@ go 1.18
 require (
 	github.com/CosmWasm/wasmd v0.29.2-osmo-v13
 	github.com/cosmos/cosmos-proto v1.0.0-alpha8
-	github.com/cosmos/cosmos-sdk v0.45.11
+	github.com/cosmos/cosmos-sdk v0.46.6
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply go.mod:8]: Oh no, it doesn't actually update the SDK either (as its replaced by the replace tag lower)

Our proto tooling just changes this :/

File: proto/osmosis/downtime-detector/v1beta1/downtime_duration.proto (hunk 3 above)
Comment (ValarDragon) [reply proto/osmosis/downtime-detector/v1beta1/downtime_duration.proto:12]: Yeah exactly! (Plus this is a proto file, which I don't think supports iota?)

File: x/downtime-detector/abci.go (hunk 4 above)
Comment (ValarDragon) [reply x/downtime-detector/abci.go:18]: Good point, will change to do that

File: x/superfluid/client/cli/query_test.go (hunk 6 above)
Comment (ValarDragon) [reply x/superfluid/client/cli/query_test.go:94]: Only insofar as I had to change the `func (s *KeeperTestHelper) StateNotAltered() {` line in apptesting because it was designed incorrectly (as triggered by this module). But fixing that, led to showing this superfluid test was designed with an error as well.

File: app/modules.go [hunk 8]
```diff
@@ -175,6 +179,7 @@ func orderBeginBlockers(allModuleNames []string) []string {
 	// IBChost came after staking, before superfluid.
 	// TODO: Come back and delete this line after testing the base change.
 	ord.Sequence(stakingtypes.ModuleName, ibchost.ModuleName, superfluidtypes.ModuleName)
+	// We leave downtime-detector un-constrained.
```
Comment (czarcas7ic) [inline app/modules.go:182]: What do we mean by un-constrained here?

File: x/downtime-detector/genesis.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	k.setGenDowntimes(ctx, gen.Downtimes)
+}
+
+func (k *Keeper) setGenDowntimes(ctx sdk.Context, genDowntimes []types.GenesisDowntimeEntry) {
+	for _, downtime := range genDowntimes {
+		k.StoreLastDowntimeOfLength(ctx, downtime.Duration, downtime.LastDowntime)
+	}
+}
+
+// ExportGenesis returns the twap module's exported genesis.
```
Comment (czarcas7ic) [inline x/downtime-detector/genesis.go:23]:
```suggestion
// ExportGenesis returns the downtime detector module's exported genesis.
```

File: x/downtime-detector/genesis.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
+package downtimedetector
```
Comment (czarcas7ic) [inline x/downtime-detector/genesis.go:1]: I believe we still need to add tests for these methods for this module. Its okay if we don't address it here, but should make an issue if not.

Edit: Ah looks like you did this in keeper_test.go. Should we move them to a genesis_test.go file?

File: x/downtime-detector/types/constants.go [hunk 11]
```diff
@@ -1,9 +1,81 @@
...
+
+	"github.com/tidwall/btree"
+)
+
 const (
 	ModuleName = "downtime-detector"
 	StoreKey   = ModuleName
-	RouterKey  = ModuleName
+	// must be alphanumeric
+	RouterKey = "downtimedetector"
```
Comment (czarcas7ic) [inline x/downtime-detector/types/constants.go:15]: I think we made concentratedliquidity just not use the "-" in the module name to prevent the diff here, might prevent some confusion down the road.

File: x/downtime-detector/types/keys.go [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package types
+
+import fmt "fmt"
+
+// There are few of these keys, so we don't concern ourselves with small key names.
+var (
+	lastBlockTimestampKey      = []byte("last_block_timestamp")
+	lastDowntimeOfLengthPrefix = "last_downtime_of_length/%s"
```
Comment (czarcas7ic) [inline x/downtime-detector/types/keys.go:8]: Was the %s supposed to be in the var itself?

**Review by czarcas7ic — APPROVED** [review]
Minor comments 👍

File: proto/osmosis/downtime-detector/v1beta1/downtime_duration.proto (hunk 3 above)
Comment (nicolaslara) [reply proto/osmosis/downtime-detector/v1beta1/downtime_duration.proto:12]: lol, right! didn't notice this was in the proto

File: x/downtime-detector/types/keys.go (hunk 12 above)
Comment (nicolaslara) [reply x/downtime-detector/types/keys.go:8]: var itself?

Comment (ValarDragon) [reply x/downtime-detector/types/keys.go:8]: Its used in a fmt.Sprintf call later

Comment (czarcas7ic) [reply x/downtime-detector/types/keys.go:8]: Ah, I have never seen it used this way but it makes sense, thanks :)

File: app/modules.go (hunk 8 above)
Comment (ValarDragon) [reply app/modules.go:182]: It can go anywhere in the sequence of begin block runs

File: x/downtime-detector/types/constants.go (hunk 11 above)
Comment (ValarDragon) [reply x/downtime-detector/types/constants.go:15]: good idea

File: x/downtime-detector/genesis.go (hunk 10 above)
Comment (ValarDragon) [reply x/downtime-detector/genesis.go:1]: did it
